src/platforms/esp32/components/avm_builtins/uart_driver.c

```diff
@@ -305,7 +308,7 @@ static void uart_driver_do_read(Context *ctx, term msg)
         term_put_tuple_element(error_tuple, 1, ealready);
 
         term result_tuple = term_alloc_tuple(2, ctx);
-        term_put_tuple_element(result_tuple, 0, ref);
+        term_put_tuple_element(result_tuple, 0, term_from_ref_ticks(ref_ticks, ctx));
```

Comment:
term_from_ref_ticks allocs a ref on the heap, so it might cause a GC. Did you book the memory for that in advance?

Reply:
Yeah good point -- definitely made a mistake assuming the GC might be called.

Why aren't you calling `memory_ensure_free` before doing the `term_alloc_tuple`?  I realize we do have a buffer of something like 8 words in the heap we can exploit with a return tuple.  Was your intention to assume there was enough free space in the heap in order to allocate the return tuple?

And is that a fair assumption, or should we be checking for free space before creating the return tuple?

Reply:
Added REF_SIZE to call to `memory_ensure_free`